readthedocs/core/views/hooks.py

```diff
@@ -100,7 +101,17 @@ def sync_versions(project):
         if not version:
             log.info('Unable to sync from %s version', version_identifier)
             return None
-        sync_repository_task.delay(version.pk)
+        task = chain(
+            sync_repository_task.signature(
+                args=(version.pk,),
+                immutable=True,
+            ),
+            clean_build_task.signature(
```

Comment:
I have a doubt here.

I understand that when we receive a webhook this task is called to sync the versions. So, if the `default_version` is triggered for a build as well, it could happen that this `clean_build_task` removes the files while the build is still building.

In other words, since calling `sync_versions` may trigger a build for `stable` version, I think that there is one case where this could break. Consider this scenario:

1. the project `default_version` is set as `stable`
1. we receive a webhook that calls `sync_versions` and it triggers a build for `stable`
1. the build for the `stable` version starts in the same builder
1. `clean_build_task` is called with `stable` version because it was the `default_version` of the project
1. files for `stable` version are removed from the builder while it was building

I understand that if we merge #5695 as well (lock the whole build), step number 5) won't happen because `clean_build_task` tries to acquire the lock at this line: https://github.com/rtfd/readthedocs.org/pull/5680/files/0a6445edb9621256f28f673e1a9ae8d4a3831704..b613bbb98f99fb507fbe84826d8fa90de2db4358#diff-b9399e1d3499066c5564f98a620e8881R1434

Please, double check this and confirm if I'm right or wrong. Although, it seems to be a very improbable case :)

Reply:
Presumably we should only be cleaning up the files for the specific version we're building?

This is another vote for single build per builder in my mind. We're wasting a lot of time worrying about this.

Reply:
I'm acquiring a lock for the clean build task in 

https://github.com/rtfd/readthedocs.org/pull/5680/files#diff-b9399e1d3499066c5564f98a620e8881R1434

And skipping the deletion if there is another task with a lock. That avoids the problem of deleting stuff if we have another task expecting files to be there.

I only need to update the code acquire a lock in the sync_versions task after #5695, that way clean_build will skip the deletion.

Reply:
> Presumably we should only be cleaning up the files for the specific version we're building?

We are doing that

Reply:
> And skipping the deletion if there is another task with a lock. That avoids the problem of deleting stuff if we have another task expecting files to be there.

Since we are not locking the whole build, this can still happen in between the locks, right?